job-control: avoid kill race

justinmk opened the pull request:
see https://github.com/neovim/neovim/issues/8269

@oni-link @bfredl @jamessan @blueyed  sanity-check appreciated.

children_kill_cb() is racey. One obvious problem is that process_close_handles() is *queued* by on_process_exit(), so when children_kill_cb() is invoked, the dead process might still be in the
`loop->children` list.  If the OS already reclaimed the dead PID, Nvim may try to SIGKILL it.

Avoid that by checking `proc->status`.

Vim doesn't have this problem because it doesn't attempt to kill processes that ignored SIGTERM after a timeout. (Future note: maybe we should not attempt this, too ...)

justinmk committed:
loop: remove `children_stop_requests`

It serves no purpose because process_stop() is already guarded by
`proc->stopped_time`.

justinmk committed:
job-control: mitigate process-kill race

children_kill_cb() is racey. One obvious problem is that
process_close_handles() is *queued* by on_process_exit(), so when
children_kill_cb() is invoked, the dead process might still be in the
`loop->children` list.  If the OS already reclaimed the dead PID, Nvim
may try to SIGKILL it.

Avoid that by checking `proc->status`.

Vim doesn't have this problem because it doesn't attempt to kill
processes that ignored SIGTERM after a timeout.

closes #8269

justinmk commented on src/nvim/event/process.c:
```diff
@@ -228,29 +228,26 @@ void process_stop(Process *proc) FUNC_ATTR_NONNULL_ALL
   }
 
   Loop *loop = proc->loop;
-  if (!loop->children_stop_requests++) {
```
Small cleanup: I removed `loop->children_stop_requests` because it seems useless.  It will always be 0 here, because `process_stop()` is guarded by `proc->stopped_time`.

justinmk commented on src/nvim/event/process.c:
```diff
@@ -228,29 +228,26 @@ void process_stop(Process *proc) FUNC_ATTR_NONNULL_ALL
     Process *proc = (*current)->data;
-    if (!proc->stopped_time) {
+    bool exited = (proc->status >= 0);
+    if (exited || !proc->stopped_time) {
```
This hopefully avoids the race in #8269

bfredl commented:
Looks sane to me on a quick glance

oni-link commented on src/nvim/event/process.c:
```diff
@@ -383,10 +380,10 @@ static void process_close_handles(void **argv)
-      && !--loop->children_stop_requests) {
-    // Stop the timer if no more stop requests are pending
-    DLOG("Stopping process kill timer");
+  ILOG("exited: pid=%d status=%d stoptime=%" PRId64, proc->pid,
```
`proc->stopped_time` is an uint64_t, so `PRIu64` instead of `PRId64`?

justinmk committed:
job-control: one-shot timer instead of repeating

Before f31c26f1afb5 the timer was used to try SIGTERM *and* SIGKILL, so
a repeating timer was needed.  After f31c26f1afb5 process_stop() sends
SIGTERM immediately, and the timer only sends SIGKILL.

So we don't need a repeating timer.
- Simplifies the logic: don't need to call uv_timer_stop() explicitly.
- Avoids a problem: if process_stop() is called more than once in the
  2-second window, the first on_process_exit() would call
  uv_timer_stop() which stops the timer for all stopped processes.

justinmk commented:
> What if we already have a process that is waiting to be killed, but the next ending process stops the kill-timer now?

@oni-link The last commit changes the timer to be non-repeating, and `uv_timer_stop` is never called.

oni-link commented on src/nvim/event/process.c:
```diff
@@ -227,11 +228,10 @@ void process_stop(Process *proc) FUNC_ATTR_NONNULL_ALL
-  uv_timer_start(&loop->children_kill_timer, children_kill_cb,
-                 KILL_TIMEOUT_MS, KILL_TIMEOUT_MS);
+  uv_timer_start(&proc->loop->children_kill_timer, children_kill_cb,
+                 KILL_TIMEOUT_MS, 0);
```
Is the process loop here also the main loop? If so all processes use the same timer and a sequence of stopped processes could reset the timer without ever being killed.

justinmk replied:
Based on libuv docs and my testing, each uv_timer_start call resets the timer. 

And the timer callback always iterates the processes. 

So I think this is fine?

oni-link commented on src/nvim/event/process.c:
```diff
@@ -227,30 +228,26 @@ void process_stop(Process *proc) FUNC_ATTR_NONNULL_ALL
     uint64_t elapsed = (now - proc->stopped_time) / 1000000 + 1;
-
     if (elapsed >= KILL_TIMEOUT_MS) {
       int sig = proc->type == kProcessTypePty && elapsed < KILL_TIMEOUT_MS * 2
```
Should this code be changed so that always a `SIGKILL` is send?
* Calling `process_stop()` for a pty that ignores `SIGTERM` would probably never on its own reach an elapsed time of `2*KILL_TIMEOUT_MS`. The `pty` would have to wait until another process is stoped so that the kill timer is restarted.
* If the timer fires too early no signal could be send and the process would again had to wait for another stopped process.

justinmk replied:
@oni-link Good catch again. @blueyed was there a specific reason we need to try SIGTERM before SIGKILL, if a PTY process didn't terminate after SIGHUP?

jamessan replied:
Is not uncommon for there to be special handling of `SIGHUP`, e.g. to reload config or show progress.  `SIGTERM`, on the other hand, is unlikely to have an overloaded meaning.  Therefore issuing `SIGTERM` still gives an opportunity for graceful shutdown before the sledgehammer of `SIGKILL`.

justinmk committed:
job-control: children_kill_cb(): do not check elapsed time

1. Don't check elapsed time in children_kill_cb(), it's already implied
   by the start-time of the timer itself.
2. Restart timer from children_kill_cb() for PTY jobs, to send SIGKILL
   after SIGTERM. There is an edge case where SIGKILL might follow
   SIGTERM too quickly, if jobstop() is called near the 2-second timer
   window.  But this edge case is not worth code complication.

justinmk commented:
>> was there a specific reason we need to try SIGTERM before SIGKILL, if a PTY process didn't terminate after SIGHUP?
>
> SIGTERM still gives an opportunity for graceful shutdown before the sledgehammer of SIGKILL.

@jamessan Ok, https://github.com/neovim/neovim/pull/8273/commits/6cb2be8acdb02aac65e6df50b19dab5e8108f8ae fixes that mostly. Though it still has an (unlikely) edge case where it's possible that SIGKILL might follow SIGTERM too quickly.

oni-link commented:
@justinmk, in what situation can a `SIGKILL` follow a `SIGTERM` too quickly for ptys?

justinmk commented:
@oni-link good call :) I was thinking of the situation where "elapsed time" was used as the guard. But now that's removed.
